File: luiz/ba2c.py
#! /usr/bin/env python


def kmer_prob(kmer, matrix):
    prob = 1.0
    pos = dict(zip("ACGT", range(4)))
    for i, nt in enumerate(kmer):
        prob *= matrix[pos[nt]][i]
    return prob


def profile_most_probable(text, k, matrix):
    most_probable_kmer = None
    max_prob = 0
    for i in range(len(text) - k + 1):
        kmer = text[i : i + k]
        prob = kmer_prob(kmer, matrix)
        if most_probable_kmer is None:
            most_probable_kmer = kmer
            max_prob = prob
        elif prob > max_prob:
            most_probable_kmer = kmer
            max_prob = prob

    return most_probable_kmer


assert (
    profile_most_probable(
        "ACCTGTTTATTGCCTAAGTTCCGAACAAACCCAATATAGCCCGAGGGCCT",
        5,
        [
            [0.2, 0.2, 0.3, 0.2, 0.3],
            [0.4, 0.3, 0.1, 0.5, 0.1],
            [0.3, 0.3, 0.5, 0.2, 0.4],
            [0.1, 0.2, 0.1, 0.1, 0.2],
        ],
    )
    == "CCGAG"
)


if __name__ == "__main__":
    with open("data/rosalind_ba2c.txt", "r") as dataset:
        text = dataset.readline().rstrip()
        k = int(dataset.readline().rstrip())
        matrix = []
        for s in dataset.readlines():
            matrix.append([float(f) for f in s.rstrip().split()])
    print(profile_most_probable(text, k, matrix))

File: luiz/src/bin/ba2c.rs
use std::env;
use std::error::Error;
use std::fs;
use std::str;

type Matrix = Vec<Vec<f64>>;

pub fn kmer_prob(kmer: &[u8], matrix: &Matrix) -> f64 {
    let mut prob = 1.;
    for (i, nt) in kmer.iter().enumerate() {
        let pos = match nt {
            b'A' => 0,
            b'C' => 1,
            b'G' => 2,
            b'T' => 3,
            _ => unimplemented!(),
        };
        prob = prob * matrix[pos][i];
    }

    prob
}

pub fn profile_most_probable(text: &str, k: usize, matrix: &Matrix) -> String {
    let mut most_probable_kmer = text[0..k].as_bytes();
    let mut max_prob = kmer_prob(most_probable_kmer, matrix);

    for kmer in text[1..].as_bytes().windows(k) {
        let prob = kmer_prob(kmer, matrix);
        if prob > max_prob {
            max_prob = prob;
            most_probable_kmer = kmer
        }
    }

    String::from_utf8_lossy(most_probable_kmer).into()
}

fn main() -> Result<(), Box<dyn Error>> {
    let input: String = env::args()
        .nth(1)
        .unwrap_or("data/rosalind_ba2c.txt".into());
    let data = fs::read_to_string(input)?;
    let mut lines = data.lines();

    let text: String = lines.next().unwrap().into();
    let k = lines.next().unwrap().parse()?;

    let mut matrix: Matrix = vec![vec![0.0; k]; 4];

    for i in 0..4 {
        lines
            .next()
            .unwrap()
            .split(' ')
            .map(|x| x.parse().unwrap())
            .take(k)
            .enumerate()
            .map(|(j, x)| matrix[i][j] = x)
            .count();
    }

    println!("{}", profile_most_probable(&text, k, &matrix));

    Ok(())
}
